src/main/comments.js

```diff
@@ -18,66 +16,51 @@ import {
   addLeadingComment,
   addDanglingComment,
   addTrailingComment,
+  isNonEmptyArray,
 } from "../common/util.js";
+import getVisitorKeys from "./get-visitor-keys.js";
 
 const childNodesCache = new WeakMap();
-function getSortedChildNodes(node, options, resultArray) {
-  if (!node) {
-    return;
-  }
-  const { printer, locStart, locEnd } = options;
-
-  if (resultArray) {
-    if (printer.canAttachComment && printer.canAttachComment(node)) {
-      // This reverse insertion sort almost always takes constant
-      // time because we almost always (maybe always?) append the
-      // nodes in order anyway.
-      let i;
-      for (i = resultArray.length - 1; i >= 0; --i) {
-        if (
-          locStart(resultArray[i]) <= locStart(node) &&
-          locEnd(resultArray[i]) <= locEnd(node)
-        ) {
-          break;
-        }
-      }
-      resultArray.splice(i + 1, 0, node);
-      return;
-    }
-  } else if (childNodesCache.has(node)) {
+function getSortedChildNodes(node, options) {
+  if (childNodesCache.has(node)) {
     return childNodesCache.get(node);
   }
 
-  const childNodes =
-    (printer.getCommentChildNodes &&
-      printer.getCommentChildNodes(node, options)) ||
-    (typeof node === "object" &&
-      Object.entries(node)
-        .filter(
-          ([key]) =>
-            key !== "enclosingNode" &&
-            key !== "precedingNode" &&
-            key !== "followingNode" &&
-            key !== "tokens" &&
-            key !== "comments" &&
-            key !== "parent"
-        )
-        .map(([, value]) => value));
-
-  if (!childNodes) {
-    return;
-  }
+  const {
+    printer: {
+      getCommentChildNodes,
+      canAttachComment,
+      getVisitorKeys: printerGetVisitorKeys,
+    },
+    locStart,
+    locEnd,
+  } = options;
 
-  if (!resultArray) {
-    resultArray = [];
-    childNodesCache.set(node, resultArray);
+  if (!canAttachComment) {
+    return [];
   }
 
-  for (const childNode of childNodes) {
-    getSortedChildNodes(childNode, options, resultArray);
-  }
+  const childNodes = (
+    getCommentChildNodes?.(node, options) ??
+    getVisitorKeys(node, printerGetVisitorKeys).flatMap((key) => node[key])
+  ).flatMap((childNode) => {
+    if (!(childNode && typeof childNode === "object")) {
+      return [];
+    }
+
+    return canAttachComment(childNode)
+      ? childNode
+      : getSortedChildNodes(childNode, options);
+  });
+
+  // Sort by `start` location first, then `end` location
+  childNodes.sort(
```

Comment:
We can't use `sort` here as it isn't stable and node can have the same loc data.

Reply:
Since v12, it's already stable https://v8.dev/features/stable-sort

Reply:
As per the ES spec, it doesn't have to be stable, and Node isn't the only engine where Prettier is supposed to work.

Reply:
Sorry, I was wrong. Didn't know that spec has changed.